### PR TITLE
Bulk download takes context and respects its cancellation

### PR DESCRIPTION
**Pull request by mhutchinson**

### CODE REVIEW

**Review by AlCutter — APPROVED** [review]
LGTM, just a suggestion for tightening the test to target more specifically what I think you're testing for.

File: clone/internal/download/batch_test.go [hunk 1]
```diff
@@ -81,6 +82,36 @@ func TestBulk(t *testing.T) {
...
+	var workers uint = 4
+	var batchSize uint = 10
+
+	fakeFetch := func(start uint64, leaves [][]byte) error {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go Bulk(ctx, first, fakeFetch, workers, batchSize, leafc, errc)
+	cancel()
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply clone/internal/download/batch_test.go]: You get a lint warning for risky behaviour of leaking contexts. However, this was fixed by calling cancel again in the main path. Good suggestion.

**Review by AlCutter — APPROVED** [review]

File: clone/internal/download/batch_test.go [hunk 2]
```diff
@@ -81,6 +82,39 @@ func TestBulk(t *testing.T) {
...
+			}
+			continue
+		case <-errc:
+		}
+		break
+	}
+	if seen == 1000 {
+		t.Error("Expected cancellation to prevent all leaves being read")
+	}
+	cancel()
```
Comment (AlCutter) [inline clone/internal/download/batch_test.go:115]: Could also `defer cancel()` on L96?

Comment (mhutchinson) [reply clone/internal/download/batch_test.go]: Better.

**Pull request merged**